rust/xaynet-mobile/src/ffi/participant.rs

```diff
@@ -345,3 +362,143 @@ pub unsafe extern "C" fn xaynet_ffi_participant_set_model(
         ERR_SETMODEL_MODEL
     }
 }
+
+/// Return the latest global model from the coordinator.
+///
+/// - `buffer` is the array in which the global model should be copied.
+/// - `data_type` specifies the type of the model weights (see [`DataType`]). The C header
+///   file generated by this crate provides an enum corresponding to the parameters: `DataType`.
+/// - `len` is the number of weights the model has
+///
+/// # Return Value
+///
+/// - [`OK`] if the model is set successfully
+/// - [`ERR_NULLPTR`] if `participant` or the `buffer` is NULL
+/// - [`GLOBAL_MODEL_NONE`] if no model exists
+/// - [`ERR_GLOBALMODEL_IO`] if the communication with the coordinator
+/// - [`ERR_GLOBALMODEL_DATATYPE`] if the datatype is invalid
+/// - [`ERR_GLOBALMODEL_LEN`] if the length of the buffer does not match the length of the model
+/// - [`ERR_GLOBALMODEL_CONVERT`] if the conversion of the model failed
+///
+/// # Note
+///
+///    It is **not** guaranteed, that the model configuration returned by
+///    [`xaynet_ffi_participant_model_config`] corresponds to the configuration of
+///    the global model. This means that the global model can have a different length / data type
+///    than it is defined in model configuration. That both model configurations are the same is
+///    only guaranteed if the model config **never** changes on the coordinator side.
+///
+/// # Safety
+///
+/// 1. When calling this method, you have to ensure that *either* the pointer is NULL
+///    *or* all of the following is true:
+///    - The pointer must be properly [aligned].
+///    - It must be "dereferencable" in the sense defined in the [`::std::ptr`] module
+///      documentation.
+/// 2. If `len` or `data_type` do not match the model in `buffer`, this method will
+///    result in a buffer over-read.
+#[no_mangle]
+pub unsafe extern "C" fn xaynet_ffi_participant_global_model(
```

Comment:
So when the coordinator returns the model config along with the model, the idea is that this function will look like this, correct?

```rust
pub unsafe extern "C" fn xaynet_ffi_participant_global_model(participant: *mut Participant, *mut ModelConfig) -> *const ByteBuffer
{
  // ...
}
```
Where `*ModelConfig` is populated based on the model config sent by the coordinator, and `ByteBuffer` contains the raw model.

Reply:
My idea was that we have something like this 
```rust
pub unsafe extern "C" fn xaynet_ffi_participant_global_model(participant: *mut Participant) -> *const ByteBuffer
{
  // p.global_model() 
  // would return a struct which contains the model as well as the data type.
  // Note that we have not currently linked the data type to the model, we just send the model

  GlobalModel {
  	model: Model,
  	type: DataType
  }

  // then we can convert the model in this function and write it in a ByteBuffer 
  let model = match global_model.type { ... }

  return GlobalModelAsCRep {
  	model: ByteBuffer,
  	type: DataType
  }
}
```

but I have already noticed a flaw in my idea
we can't use `ByteBuffer` here because it can only holds an `u8` array. The data type of a model however can be `f32`, `f64`, `i32` or `i64`. So I don't know what the reutrn type should be. 
what we maybe can do is a create a buffer that holds a c_void pointer instead of `u8` but I'm not sure if this is possible  if we can implement this in a safe way.